File: src/tools/read.zig
const std = @import("std");
const core = @import("phoenix_core");

const default_limit: usize = 2000;
const max_bytes: usize = 512 * 1024;

const image_exts = [_][]const u8{ ".jpg", ".jpeg", ".png", ".gif", ".webp" };

fn errResult(allocator: std.mem.Allocator, comptime fmt: []const u8, args: anytype) anyerror!core.tool.ToolResult {
    const msg = try std.fmt.allocPrint(allocator, fmt, args);
    return .{ .output = msg, .is_error = true };
}

fn isImagePath(path: []const u8) bool {
    var i: usize = path.len;
    while (i > 0) : (i -= 1) {
        if (path[i - 1] == '.') {
            const ext = path[i - 1 ..];
            for (image_exts) |e| {
                if (std.ascii.eqlIgnoreCase(ext, e)) return true;
            }
            return false;
        }
        if (path[i - 1] == '/') return false;
    }
    return false;
}

fn invoke(self: *const core.Tool, io: std.Io, args: []const u8, allocator: std.mem.Allocator) anyerror!core.tool.ToolResult {
    _ = self;

    var parsed = std.json.parseFromSlice(std.json.Value, allocator, args, .{}) catch {
        return errResult(allocator, "read: invalid JSON args: {s}", .{args});
    };
    defer parsed.deinit();
    if (parsed.value != .object) return errResult(allocator, "read: args must be a JSON object", .{});
    const obj = parsed.value.object;

    const path_v = obj.get("path") orelse return errResult(allocator, "read: missing required field 'path'", .{});
    if (path_v != .string) return errResult(allocator, "read: 'path' must be a string", .{});
    const path = path_v.string;

    if (isImagePath(path)) {
        const msg = try std.fmt.allocPrint(allocator, "read: image attachments not yet supported (path={s})", .{path});
        return .{ .output = msg, .is_error = true };
    }

    var offset: usize = 1;
    if (obj.get("offset")) |v| {
        if (v != .integer or v.integer < 1) return errResult(allocator, "read: 'offset' must be a positive integer (1-indexed)", .{});
        offset = @intCast(v.integer);
    }

    var limit: usize = default_limit;
    if (obj.get("limit")) |v| {
        if (v != .integer or v.integer < 1) return errResult(allocator, "read: 'limit' must be a positive integer", .{});
        limit = @intCast(v.integer);
    }

    const all = std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(max_bytes)) catch |err| {
        return errResult(allocator, "read: cannot read {s}: {s}", .{ path, @errorName(err) });
    };
    defer allocator.free(all);

    // Find byte offsets of the (offset)th and (offset+limit)th newlines
    // and slice between them. This preserves trailing-newline semantics
    // exactly (no synthetic newlines added or dropped).
    var start: usize = 0;
    var skipped: usize = 0;
    while (skipped + 1 < offset) {
        const nl = std.mem.indexOfScalarPos(u8, all, start, '\n') orelse {
            // File has fewer lines than `offset` — return empty.
            const owned = try allocator.dupe(u8, "");
            return .{ .output = owned };
        };
        start = nl + 1;
        skipped += 1;
    }

    var end: usize = start;
    var taken: usize = 0;
    while (taken < limit) {
        const nl = std.mem.indexOfScalarPos(u8, all, end, '\n') orelse {
            end = all.len;
            break;
        };
        end = nl + 1;
        taken += 1;
    }

    const slice = all[start..end];
    const owned = try allocator.dupe(u8, slice);
    return .{ .output = owned };
}

pub const read_tool: core.Tool = .{
    .name = "read",
    .description = "Read the contents of a file. Supports text files. For text files, defaults to first 2000 lines. Use offset/limit for large files. Image files (jpg, png, gif, webp) are not yet supported.",
    .schema =
    \\{"type":"object","properties":{"path":{"type":"string","description":"Path to the file to read (relative or absolute)"},"offset":{"type":"integer","minimum":1,"description":"Line number to start reading from (1-indexed)"},"limit":{"type":"integer","minimum":1,"description":"Maximum number of lines to read"}},"required":["path"]}
    ,
    .max_output_bytes = max_bytes,
    .invokeFn = invoke,
};

// ---- Tests ----

test "read default returns whole small file" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_read_default.txt";
    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = tmp_path, .data = "alpha\nbeta\ngamma\n" });
    defer std.Io.Dir.cwd().deleteFile(io, tmp_path) catch {};

    const args =
        \\{"path":"phoenix_test_read_default.txt"}
    ;
    const r = try read_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expectEqualStrings("alpha\nbeta\ngamma\n", r.output);
}

test "read offset and limit slice" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const tmp_path = "phoenix_test_read_slice.txt";
    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = tmp_path, .data = "one\ntwo\nthree\nfour\nfive\n" });
    defer std.Io.Dir.cwd().deleteFile(io, tmp_path) catch {};

    const args =
        \\{"path":"phoenix_test_read_slice.txt","offset":2,"limit":2}
    ;
    const r = try read_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expectEqualStrings("two\nthree\n", r.output);
}

test "read image extension is rejected" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const args =
        \\{"path":"foo/bar.PNG"}
    ;
    const r = try read_tool.invoke(io, args, a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "image attachments not yet supported") != null);
}

test "read missing path is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const r = try read_tool.invoke(io, "{}", a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "missing required field 'path'") != null);
}

test "read invalid JSON is error" {
    const a = std.testing.allocator;
    const io = std.testing.io;
    const r = try read_tool.invoke(io, "not json", a);
    defer a.free(r.output);
    try std.testing.expect(r.is_error);
    try std.testing.expect(std.mem.indexOf(u8, r.output, "invalid JSON") != null);
}

File: src/tools/read.rs
use async_trait::async_trait;
use serde_json::{Value, json};

use super::traits::{InputRequester, Tool, ToolError, ToolResult, ToolSchema};

/// Default number of lines returned when `limit` is not specified.
const DEFAULT_LIMIT: usize = 2000;

/// Maximum file size we will read (512 KB).
const MAX_BYTES: usize = 512 * 1024;

pub struct ReadTool;

#[async_trait]
impl Tool for ReadTool {
    fn schema(&self) -> ToolSchema {
        ToolSchema {
            name: "read".into(),
            description:
                "Read the contents of a file. Returns line-numbered output (cat -n style). \
                          Defaults to the first 2000 lines. Use offset/limit for large files."
                    .into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "file_path": {
                        "type": "string",
                        "description": "Absolute path to the file to read"
                    },
                    "offset": {
                        "type": "integer",
                        "minimum": 0,
                        "description": "Line number to start reading from (0-indexed, default 0)"
                    },
                    "limit": {
                        "type": "integer",
                        "minimum": 1,
                        "description": "Maximum number of lines to read (default 2000)"
                    }
                },
                "required": ["file_path"]
            }),
        }
    }

    async fn invoke(
        &self,
        args: Value,
        _input: &dyn InputRequester,
    ) -> Result<ToolResult, ToolError> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::InvalidArgs("missing required field 'file_path'".into()))?;

        let offset = args.get("offset").and_then(|v| v.as_u64()).unwrap_or(0) as usize;

        let limit = args
            .get("limit")
            .and_then(|v| v.as_u64())
            .map(|v| v as usize)
            .unwrap_or(DEFAULT_LIMIT);

        // Read the file (capped at MAX_BYTES).
        let raw = match tokio::fs::read(file_path).await {
            Ok(bytes) => {
                if bytes.len() > MAX_BYTES {
                    bytes[..MAX_BYTES].to_vec()
                } else {
                    bytes
                }
            }
            Err(e) => {
                return Err(ToolError::ExecutionFailed(format!(
                    "cannot read {file_path}: {e}"
                )));
            }
        };

        let content = String::from_utf8_lossy(&raw);
        let lines: Vec<&str> = content.lines().collect();

        // Apply offset and limit.
        let start = offset.min(lines.len());
        let end = (start + limit).min(lines.len());
        let selected = &lines[start..end];

        // Format with line numbers (1-indexed, matching `cat -n` style).
        let mut out = String::new();
        for (i, line) in selected.iter().enumerate() {
            let line_num = start + i + 1; // 1-indexed
            out.push_str(&format!("{line_num:>6}\t{line}\n"));
        }

        let truncated = raw.len() >= MAX_BYTES || end < lines.len();

        Ok(ToolResult {
            output: out,
            truncated,
            is_error: false,
            toast: None,
            widget_json: None,
        })
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tools::traits::NoopInputRequester;
    use tempfile::TempDir;

    fn write_temp(dir: &TempDir, name: &str, content: &str) -> String {
        let path = dir.path().join(name);
        std::fs::write(&path, content).unwrap();
        path.to_str().unwrap().to_string()
    }

    #[tokio::test]
    async fn read_whole_small_file() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "test.txt", "alpha\nbeta\ngamma\n");

        let result = ReadTool
            .invoke(json!({"file_path": path}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(!result.is_error);
        assert!(result.output.contains("alpha"));
        assert!(result.output.contains("beta"));
        assert!(result.output.contains("gamma"));
        // Check line numbers are present.
        assert!(result.output.contains("1\t"));
        assert!(result.output.contains("2\t"));
        assert!(result.output.contains("3\t"));
    }

    #[tokio::test]
    async fn read_with_offset_and_limit() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "slice.txt", "one\ntwo\nthree\nfour\nfive\n");

        let result = ReadTool
            .invoke(
                json!({"file_path": path, "offset": 1, "limit": 2}),
                &NoopInputRequester,
            )
            .await
            .unwrap();
        assert!(!result.is_error);
        // offset=1 means skip line 1 ("one"), start at line 2 ("two").
        assert!(result.output.contains("two"));
        assert!(result.output.contains("three"));
        assert!(!result.output.contains("\tone\n"));
        assert!(!result.output.contains("\tfour\n"));
    }

    #[tokio::test]
    async fn read_offset_beyond_file() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "short.txt", "one\ntwo\n");

        let result = ReadTool
            .invoke(
                json!({"file_path": path, "offset": 100}),
                &NoopInputRequester,
            )
            .await
            .unwrap();
        assert!(!result.is_error);
        assert_eq!(result.output, "");
    }

    #[tokio::test]
    async fn read_missing_file() {
        let result = ReadTool
            .invoke(
                json!({"file_path": "/tmp/nonexistent_phx_test_file.txt"}),
                &NoopInputRequester,
            )
            .await;
        match result {
            Err(ToolError::ExecutionFailed(msg)) => {
                assert!(msg.contains("cannot read"), "msg was: {msg}");
            }
            other => panic!("expected ExecutionFailed, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn read_missing_path_arg() {
        let result = ReadTool.invoke(json!({}), &NoopInputRequester).await;
        match result {
            Err(ToolError::InvalidArgs(msg)) => {
                assert!(msg.contains("file_path"), "msg was: {msg}");
            }
            other => panic!("expected InvalidArgs, got: {other:?}"),
        }
    }

    #[tokio::test]
    async fn read_line_numbers_are_one_indexed() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "nums.txt", "aaa\nbbb\n");

        let result = ReadTool
            .invoke(json!({"file_path": path}), &NoopInputRequester)
            .await
            .unwrap();
        let first_line = result.output.lines().next().unwrap();
        // Should start with "     1\t"
        assert!(
            first_line.trim_start().starts_with("1\t"),
            "first line was: {first_line}"
        );
    }

    #[tokio::test]
    async fn read_empty_file() {
        let dir = TempDir::new().unwrap();
        let path = write_temp(&dir, "empty.txt", "");

        let result = ReadTool
            .invoke(json!({"file_path": path}), &NoopInputRequester)
            .await
            .unwrap();
        assert!(!result.is_error);
        assert_eq!(result.output, "");
    }
}
